sw/host/opentitanlib/src/transport/cw310/mod.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{ensure, Result};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use crate::io::gpio::Gpio;
+use crate::io::spi::Target;
+use crate::io::uart::Uart;
+use crate::transport::{Capabilities, Capability, Transport, TransportError};
+
+pub mod gpio;
+pub mod spi;
+pub mod uart;
+pub mod usb;
+
+#[derive(Default)]
+struct Inner {
+    spi: Option<Rc<dyn Target>>,
+    gpio: Option<Rc<dyn Gpio>>,
+    uart: HashMap<u32, Rc<dyn Uart>>,
+}
+
+pub struct CW310 {
+    device: Rc<RefCell<usb::Backend>>,
+    inner: RefCell<Inner>,
+}
+
+impl CW310 {
+    // Pins needed for reset & bootstrap on the CW310 board.
+    const PIN_SRST: &'static str = "USB_A18";
+    const PIN_BOOTSTRAP: &'static str = "USB_A16";
+    // Pins needed for SPI on the CW310 board.
+    const PIN_CLK: &'static str = "USB_SPI_SCK";
+    const PIN_SDI: &'static str = "USB_SPI_COPI";
+    const PIN_SDO: &'static str = "USB_SPI_CIPO";
+    const PIN_CS: &'static str = "USB_SPI_CS";
+    const PIN_JTAG: &'static str = "USB_A19";
+
+    pub fn new(
+        usb_vid: Option<u16>,
+        usb_pid: Option<u16>,
+        usb_serial: Option<String>,
+    ) -> Result<Self> {
+        let board = CW310 {
+            device: Rc::new(RefCell::new(usb::Backend::new(
+                usb_vid, usb_pid, usb_serial,
+            )?)),
+            inner: RefCell::default(),
+        };
+        board.init_direction()?;
+        Ok(board)
+    }
+
+    // Initialize the IO direction of some basic pins on the board.
+    fn init_direction(&self) -> Result<()> {
+        let device = self.device.borrow();
+        device.pin_set_output(Self::PIN_SRST, true)?;
+        device.pin_set_output(Self::PIN_JTAG, true)?;
+        device.pin_set_output(Self::PIN_BOOTSTRAP, true)?;
+        Ok(())
+    }
+}
+
+impl Transport for CW310 {
+    fn capabilities(&self) -> Capabilities {
+        Capabilities::new(Capability::SPI | Capability::GPIO | Capability::UART)
+    }
+
+    fn uart(&self, instance: u32) -> Result<Rc<dyn Uart>> {
+        let mut inner = self.inner.borrow_mut();
+        // I'd prefer to follow clippy's advice here, but I need the error-return
+        // behavior of the `?` operator when opening the uart.
+        #[allow(clippy::map_entry)]
+        if !inner.uart.contains_key(&instance) {
+            inner.uart.insert(
+                instance,
+                Rc::new(uart::CW310Uart::open(Rc::clone(&self.device), instance)?),
+            );
+        }
+        Ok(Rc::clone(inner.uart.get(&instance).unwrap()))
+    }
+
+    fn gpio(&self) -> Result<Rc<dyn Gpio>> {
+        let mut inner = self.inner.borrow_mut();
+        if inner.gpio.is_none() {
```

Comment:
Shame there isn't a get_or_try_insert. =/

You could do something similar to what I did above with the map but it's probably not worth it.

Reply:
Indeed.

Or a feature for closures to capture short-circuit Result<T,E> return path and short circuit out to the next appropriate enclosing scope... although I imagine implementing such a thing is far trickier because of whatever intermediate state the function calling the closure might have.